### PR TITLE
Header processing framework

### PR DESCRIPTION
**Pull request by tnawathe21**
## Header Processing System: High level structure

Looking for general feedback on this proposed parsing framework!

### Parsing
- Http filter splits yaml config into operations (eg. [set-header key1 val1, set-path path1, set-header key2 val2]
- Each operation is then sent to its respective HeaderProcessor instance
- HeaderProcessor will parse out the key/value or the path (based on the type of operation) as well as the condition
- Condition is sent to ConditionProcessor which will parse out each acl, send each acl to an AclProcessor to evaluate, and evaluate the whole condition

### Evaluating
- (to be implemented) Http filter will store all operations as HeaderProcessor instances -- one vector of instances for the request side and a separate vector for the response side
- (to be implemented) in encodeHeaders/decodeHeaders, Http filter will iterate through the respective vector and execute the operation
- HeaderProcessor instances will perform the operation based on the parsed tokens

Note: build fails because some parameters are unused

### CODE REVIEW

**Review by rob05c — COMMENTED** [review]
This seems like a good approach to me.

At the end of the day, I think for any implementation:
1. the parsed expressions will be an ad-hoc AST
2. the evaluation will be a State Machine

There are a lot of ways to do a State Machine in C++, I'm not sure any one is better than the others. Classes are good, and the inheritance here makes sense.

Likewise, an array of Processors seems like a perfectly fine way to do the "AST" of parsed config.

I think some things will be necessary in the future: 

Each of these `Processor` is going to need state from the previous processors, from ACLs (which are how HAProxy does variables). So each Processor will need to take a "WorldState" or "AclData" object at evaluation time, along with the current HTTP request.

And the list of Processors that run in order, like you mention.

But nothing here precludes those.

TLDR the general approach looks good to me. I don't see any fundamental issues or anything.

**Comment by rob05c** [comment]
> ACLs (which are how HAProxy does variables)

Incidentally, what HAProxy calls "ACLs" aren't what anybody else means by "ACL," they're really just variables.

If the goal is to be compatible with HAProxy config, that's fine. But if not, I'd vote we name the directive `set-variable` or something more obvious.

I think we're going to want to extend the syntax anyway; if nothing else, `set-metadata` and `get-metadata` will be invaluable for composing with other filters, logging, etc
